src/stream.nim

```diff
@@ -0,0 +1,53 @@
+ import pkg/[chronos,
+            chronos/apps/http/httpserver,
+            chronos/timer,
+            chronicles,
+            httputils]
+import strutils
+
+const
+  HttpHeadersTimeout = timer.seconds(120) # timeout for receiving headers (120 sec)
+  HeaderSep = @[byte('\c'), byte('\L'), byte('\c'), byte('\L')]
+  MaxHttpHeadersSize = 8192       # maximum size of HTTP headers in octets
+
+
+proc readHeaders*(rstream: AsyncStreamReader): Future[seq[byte]] {.async.} =
+  var buffer = newSeq[byte](MaxHttpHeadersSize)
+  var error = false
+  try:
+    let hlenfut = rstream.readUntil(addr buffer[0], MaxHttpHeadersSize,
+        sep = HeaderSep)
+    let ores = await withTimeout(hlenfut, HttpHeadersTimeout)
+    if not ores:
+      # Timeout
+      debug "Timeout expired while receiving headers",
+            address = rstream.tsource.remoteAddress()
+      error = true
+    else:
+      let hlen = hlenfut.read()
+      buffer.setLen(hlen)
+  except TransportLimitError:
+    # size of headers exceeds `MaxHttpHeadersSize`
+    debug "Maximum size of headers limit reached",
+          address = rstream.tsource.remoteAddress()
+    error = true
+  except TransportIncompleteError:
+    # remote peer disconnected
+    debug "Remote peer disconnected", address = rstream.tsource.remoteAddress()
+    error = true
+  except TransportOsError as exc:
+    debug "Problems with networking", address = rstream.tsource.remoteAddress(),
+          error = exc.msg
+    error = true
+
+  if error:
+    buffer.setLen(0)
+  return buffer
+
+proc closeWait*(wsStream : AsyncStream): Future[void] {.async.} =
+  if not wsStream.writer.tsource.closed():
```

Comment:
This is a little bit wrong. You assume here that `AsyncStream` is wrapper around `StreamTransport`, but its not always true.
When you get `AsyncStreamReader`, `AsyncStreamWriter` from HttpServer interface you can get wrappers only if connection is non-secure (plain HTTP), but if connection is secure `AsyncStreamReader/Writer` would be actually `TLSStreamReader/Writer` and corresponding `tsource` will be `nil`, while corresponding `rsource` will be `AsyncStreamReader/Writer` wrappers.

Reply:
@arijitAD why not just do `wsStream.closeWait()`?

Reply:
Or better yet, just leave this method out since the strems already expose this method and there is no reason for another level of indirection.

Reply:
@cheatfate, why doesn't chronos expose a sane `closeWait()` version to close this streams? It's a bit akward to have to care about all the subtleties of what is set when and on which version of the stream. 

Can't we just have `closeWait()` method that works for all this versions of the streams? I tried figuring out what the correct procedure to close them is, but it's deinitely not straightforward.